resiprocate-1.8/rutil/dns/ares/ares_send.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.\"
.TH ARES_SEND 3 "25 July 1998"
.SH NAME
ares_send \- Initiate a DNS query
.SH SYNOPSIS
.nf
.B #include <ares.h>
.PP
.B typedef void (*ares_callback)(void *\fIarg\fP, int \fIstatus\fP,
.B	unsigned char *\fIabuf\fP, int \fIalen\fP)
.PP
.B
void ares_send(ares_channel \fIchannel\fP, const unsigned char *\fIqbuf\fP,
.B 	int \fIqlen\fP, ares_callback \fIcallback\fP, void *\fIarg\fP)
.fi
.SH DESCRIPTION
The
.B ares_send
function initiates a DNS query on the name service channel identified
by
.IR channel .
The parameters
.I qbuf
and
.I qlen
give the DNS query, which should already have been formatted according
to the DNS protocol.  When the query is complete or has failed, the
ares library will invoke
.IR callback .
Completion or failure of the query may happen immediately, or may
happen during a later call to
.BR ares_process (3)
or
.BR ares_destroy (3).
.PP
The callback argument
.I arg
is copied from the
.B ares_send
argument
.IR arg .
The callback argument
.I status
indicates whether the query succeeded and, if not, how it failed.  It
may have any of the following values:
.TP 19
.B ARES_SUCCESS
The query completed.
.TP 19
.B ARES_EBADQUERY
The query buffer was poorly formed (was not long enough for a DNS
header or was too long for TCP transmission).
.TP 19
.B ARES_ETIMEOUT
No name servers responded within the timeout period.
.TP 19
.B ARES_ECONNREFUSED
No name servers could be contacted.
.TP 19
.B ARES_ENOMEM
Memory was exhausted.
.TP 19
.B ARES_EDESTRUCTION
The name service channel
.I channel
is being destroyed; the query will not be completed.
.PP
If the query completed, the callback argument
.I abuf
points to a result buffer of length
.IR alen .
If the query did not complete,
.I abuf
will be NULL and
.I alen
will be 0.
.PP
Unless the flag
.B ARES_FLAG_NOCHECKRESP
was set at channel initialization time,
.B ares_send
will normally ignore responses whose questions do not match the
questions in
.IR qbuf ,
as well as responses with reply codes of
.BR SERVFAIL ,
.BR NOTIMP ,
and
.BR REFUSED .
Unlike other query functions in the ares library, however,
.B ares_send
does not inspect the header of the reply packet to determine the error
status, so a callback status of
.B ARES_SUCCESS
does not reflect as much about the response as for other query
functions.
.SH SEE ALSO
.BR ares_process (3)
.SH AUTHOR
Greg Hudson, MIT Information Systems
.br
Copyright 1998 by the Massachusetts Institute of Technology.


ChangeLog
= Release Notes v 1.8 =

==1.8.13 Changes==

==1.8.12 Changes==
* repro: enable IPv6 by default
* resip/recon: test cases: suppress running unitTests binary, requires audio hardware
* Add new macro RESIP_BIG_ENDIAN, fixes for big endian users
* rpm: add new reTurn client header locations to spec file
* resip/stack: Tuple: fix mask comparison for some platforms with 64 bit long
* Android: various fixes, add support for DNS and logging
* resip/stack: remove use of RSA_generate_key (depracated in OpenSSL)
* various fixes for potential buffer overflow (not in SIP parsing)
* TFM: finish setting up for autotools builds
* rutil: Stabilize the ABI of Poll
* resip/dum: revert fix for onNewSubscription() bug committed in 10183 (r1.8.8)

==1.8.11 Changes==
* resip/recon: MOHParkServer: add ssl to libs
* fixed incorrect placement of #ifndef RESIP_USE_STIL_STREAMS as reported by Thomas Troy
* fix for issue introduced in rev 9367 where a connection terminated event doesn't get properly dispatched to DUM - thanks to Jorge Bustamante for reporting this
* rutil: MD5 code: more checks for big endian environments (gcc __BYTE_ORDER__)
* Fixing a goof that was causing later code to operate on unaligned data as if it were aligned, which was causing the code to malfunction on many non-x86 platforms.
* No longer trying to log mNextTransmission when it is possibly NULL
* rutil: Android cross-compile: fix compile issues
* build: add initial version of configure-android.sh
* resip/stack: testSipMessage: Remove a very strange extraneous '#', scope a test
* repro: fixed CommandServer being created even though listener bind failed
* repro: fix port already in use error on linux, resulting from binding to both INADDR_ANY and in6addr_any
* repro: added support for WebAdmin and CommandServer binding to specific IP addresses
* rpm: configure repro and reTurn not to fork under systemd in Fedora (contributed by Michael Scherer)

==1.8.10 Changes==
* add missed changelog entries from 1.8.9 release
* autotools: add -lcrypto when using -lssl

==1.8.9 Changes==
* resip/stack: TcpConnection: adjust logging levels
* rutil: more specific check for Solaris, as there are now Linux builds on SPARC
* reTurn: fix bugs with port allocation logic
* applied patch from Tibor Velencei to fix issue with over active 1xx retransmissions
* reTurn: correct header installation
* treat EGAIN and EWOULDBLOCK the same
* reTurn/client: adjust paths in headers

==1.8.8 Changes==
*added VS2010 project files for MOHParkServer
*replaced deprecated bzero calls with memsets since bzero doesn't exist in Visual Studio compiler
*allow AllowEvents to be an advertised capability
*reflow, recon: Cumulative improvements to enable recon build
*Fixed a bug that would prevent onNewSubscription (client) from being called
*ensure we can send a 100 Trying for NonInviteTransaction retransmissions
*rpm: update spec for older RH/Fedora versions, add init scripts
*rpm: libtool hack to avoid rpath problems
*fixing ClientRegistration issue (client registration can end on transport switchover)
*don't assert in Tuple constructors that take printable addresses, since these addresses can come from the wire or from configuration
*change to Helper::isClientBehindNAT to ensure we don't pass hostnames that might be Via host (ie: web-rtc) to a Tuple constructor
*changed a couple more places where we should catch BaseException instead of ParseException
*widen exception catching net
**ensures we also catch ParseExceptions when retrieving transaction id's, a problem reported by Roman Romanchenko
*don't create a Tuple out of a Via hostname if it isn't an IP address
*exceptions where seen repro when handling web-rtc messages that contain hostnames in the via host
*ensure that we can accept IPv6 addresses in the Transport settings in repro.config

==1.8.7 Changes==
*doc: Update README now that we use autotools
*dum: ensured that all DUM usages/sessions DumCommandAdapter commands are using handles rather than references when saving the usage/session they apply to.
**This fixes potential crashes in the case where a DumCommandAdapter would be enqueued onto the stack's queue while a usage terminating message is sitting in the queue.
**The effect in that case would be that the saved reference to the usage would become invalid and this could cause a crash.
**Also fixed an issue with InviteSession::*NITCommand() where DumCommandAdapter commands would end up being recursively queued onto the stack.
*stack: there are places in the code that will call getHeaderName with Headers::NONE.  This value is not within the HeaderNames array bounds, so we check the type passed in to ensure it is in range so we won't seg fault
*eclipse: add Ellemtel code style definition
*autotools: add convenient configure option for pedantic stack behavior
*stack: TransactionState: discard repeated requests received over reliable transports
*rutil: logging: stop sending process name to syslog, send PID instead
*rutil: put back timestamps and mAppName when logging to Syslog (mAppName to be configurable from repro)
*repro: add LoggingInstanceName config option
*ensure that we can accept IPv6 addresses in the Transport settings in repro.config
*repro: fixes for CommonNameMappings, display meaningful error (backport r10052, r10053)
*rpm: various spec file fixes (backport r10044:10106 from main)
*travis-ci: add .travis.yml to 1.8 release branch
*rutil/test: testData: make comparison of double more tolerant (backport r10037)

==1.8.6 Changes==
*rutil: fix for crash in DNS code (when DnsStub destructor calls setPollGrp(0))
*stack: some memory leaks reported by Shypytiak Taras - some at stack destruction time
*stack: fix to Fifo to avoid calling AsyncProcessHandler in addMultiple if adding a list of 0 messages - was causing high CPU on process loops
*stack: change to ProducerFifoBuffer to avoid calling mFifo.addMultiple if mBuffer is empty on flush()

rutil/dns/ares/ares_send.c
/* Copyright 1998 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */


#include <sys/types.h>

#ifndef WIN32
#include <netinet/in.h>
#ifndef __CYGWIN__
#  include <arpa/nameser.h>
#endif
#endif

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "ares.h"
#include "ares_dns.h"
#include "ares_private.h"

void ares_send(ares_channel channel, const unsigned char *qbuf, int qlen,
	       ares_callback callback, void *arg)
{
  struct query *query;
  int i;
  time_t now;

  /* Verify that the query is at least long enough to hold the header. */
  if (qlen < HFIXEDSZ || qlen >= (1 << 16))
    {
      callback(arg, ARES_EBADQUERY, NULL, 0);
      return;
    }

  /* Allocate space for query and allocated fields. */
  query = malloc(sizeof(struct query));
  if (!query)
    {
      callback(arg, ARES_ENOMEM, NULL, 0);
      return;
    }
  query->tcpbuf = malloc(qlen + 2);
  if (!query->tcpbuf)
    {
      free(query);
      callback(arg, ARES_ENOMEM, NULL, 0);
      return;
    }
    if (channel->nservers)
    {
      query->skip_server = malloc(channel->nservers * sizeof(int));
    }
    else
    {
        query->skip_server = 0;
    }
  
    if (!query->skip_server)
    {
      free(query->tcpbuf);
      free(query);
      callback(arg, ARES_ENOMEM, NULL, 0);
      return;
    }

  /* Compute the query ID.  Start with no timeout. */
  query->qid = DNS_HEADER_QID(qbuf);
  query->timeout = 0;

  /* Form the TCP query buffer by prepending qlen (as two
   * network-order bytes) to qbuf.
   */
  query->tcpbuf[0] = (qlen >> 8) & 0xff;
  query->tcpbuf[1] = qlen & 0xff;
  memcpy(query->tcpbuf + 2, qbuf, qlen);
  query->tcplen = qlen + 2;

  /* Fill in query arguments. */
  query->qbuf = query->tcpbuf + 2;
  query->qlen = qlen;
  query->callback = callback;
  query->arg = arg;

  /* Initialize query status. */
  query->itry = 0;
  query->server = 0;
  for (i = 0; i < channel->nservers; i++)
    query->skip_server[i] = 0;
  query->using_tcp = (channel->flags & ARES_FLAG_USEVC) || qlen > PACKETSZ;
  query->error_status = ARES_ECONNREFUSED;

  /* Chain the query into this channel's query list. */
  query->next = channel->queries;
  channel->queries = query;

  /* Perform the first query action. */
  time(&now);
  ares__send_query(channel, query, now);
}
